### ISSUE TITLE
Can't install kicad-diff on windows

### ISSUE THREAD

**Issue by jorlee99** [leoheck/kdiff#1]
I am trying to install the external dependencies on a windows machine but when I go through the steps of installing kicad-diff I get the following.

After running the command `opam install plotkicadsch`

`[ERROR] The compilation of plotkicadsch failed at
        "C:\\OCaml64\\home\\Stanley\\.opam\\4.09.1+mingw64c\\bin\\dune.exe
        build -p plotkicadsch -j 5".

=== ERROR while compiling plotkicadsch.0.8.0 ==
===============================#
 context     2.0.8 | win32/x86_64 | ocaml-variants.4.09.1+min
gw64c | pinned(file://C:/Users/Stanley/Desktop/ELEC)
 path        ~/.opam/4.09.1+mingw64c/.opam-switch/build/plotk
icadsch.0.8.0
 command     C:\OCaml64\home\Stanley\.opam\4.09.1+mingw64c\bi
n\dune.exe build -p plotkicadsch -j 5
exit-code   1
env-file    ~/.opam/log/plotkicadsch-21024-42bb52.env
 output-file ~/.opam/log/plotkicadsch-21024-42bb52.out
output 
File "plotkicadsch/plotkicadsch/src/dune", line 7, characters 6-1
4:
7 |       kicadsch
         ^^^^^^^^
 Error: Library "kicadsch" not found.
Hint: try:
dune external-lib-deps --missing -p plotkicadsch -j 5 @install`

And running the command suggested by the hint I get 
`$ dune external-lib-deps --missing -p plotkicadsch -j 5 @install
Error: I don't know about package plotkicadsch (passed through -p)`

I am not very sure how to get this running on windows

**Comment by leoheck** [comment]
Check this answer from the Author of the `plotkicadsch`.
https://github.com/jnavila/plotkicadsch/issues/49#issuecomment-839178456

**Comment by jorlee99** [comment]
> Check this answer from the Author of the `plotkicadsch`.
> [jnavila/plotkicadsch#49 (comment)](https://github.com/jnavila/plotkicadsch/issues/49#issuecomment-839178456)

I think this worked thanks for the help there! I posted my other error within that thread so hopefully that will get figured out. It seems I'm also getting errors on the KiCad-Diff plugin as well. It seems simpler but I am not sure?

So I'm using my bash console to change the path using the following command
`source env.sh` 
and that seems to run just fine. But as soon as I run my `./kicad --help` command I get the following
`/usr/bin/env: ‘python3’: No such file or directory`
Am I possibly missing something basic here?

Thanks for your help!

**Comment by jorlee99** [comment]
> > Check this answer from the Author of the `plotkicadsch`.
> > [jnavila/plotkicadsch#49 (comment)](https://github.com/jnavila/plotkicadsch/issues/49#issuecomment-839178456)
> 
> I think this worked thanks for the help there! I posted my other error within that thread so hopefully that will get figured out. It seems I'm also getting errors on the KiCad-Diff plugin as well. It seems simpler but I am not sure?
> 
> So I'm using my bash console to change the path using the following command
> `source env.sh`
> and that seems to run just fine. But as soon as I run my `./kicad --help` command I get the following
> `/usr/bin/env: ‘python3’: No such file or directory` so it seems it is trying to run the python file itself? Is the PATH variable not being set?
> Am I possibly missing something basic here?
> 
> Thanks for your help!

**Comment by leoheck** [comment]
launch kicad in a nother terminal.

**Comment by jorlee99** [comment]
Ok I've made some progress in the mean time. I've gotten up to here on the windows machine. As for reference I am using the [sample project](https://jnavila.github.io/plotkicadsch/plotgitsch_usersguide.html) from plotgitsch. I get the following on my webpage

Schematic Layout:
![image](https://user-images.githubusercontent.com/43526839/118042049-6e911e00-b328-11eb-8c69-3cbf06159aed.png)

PCB Layout:
![image](https://user-images.githubusercontent.com/43526839/118042788-5bcb1900-b329-11eb-8b9a-0db1822d4811.png)

As you can see nothing is output for the schematic even though it should. As for the PCB layout it seems to take some place holder img file? I am not sure. 

I have tested this on my linux OS of the project and it seems like it works fine? The schematic works just fine I have yet to test the PCB layout of the project. Any ideas? One thing to note is that when it outputs 
`Launching /cygdrive/c/Users/Stanley/Desktop/kdiff/kicad_playground/kidiff/schematic/web//main.html`
No browser actually launches, this is probably due to the fact that I am using windows and opening it from the html file is not too much of a hassle.

**Comment by leoheck** [comment]
Delete the kidiff folder or use the -r flag when executing kdiff and test one more time.
But the projects you are showing does not have changes in the layout.

**Comment by jorlee99** [comment]
Unfortunately there was no change after removing to folder. Yes you are correct there are no layout changes, but it seems as though it can't find the file itself and uses some placeholder image?

> Delete the kidiff folder or use the -r flag when executing kdiff and test one more time.
> But the projects you are showing does not have changes in the layout.

**Comment by leoheck** [comment]
run it with the flag -V then it will show each command it is running
you can run each command manually to check if it is generating anything.

**Comment by jorlee99** [comment]
Sure

```
$ kdiff schematic/kicad_playground.kicad_pcb -V

    script_dir: /cygdrive/c/Users/Stanley/Desktop/kdiff/kdiff

     repo_root: /cygdrive/c/Users/Stanley/Desktop/kdiff/kicad_playground
   project_dir: schematic

     kicad_pcb: schematic/kicad_playground.kicad_pcb
     sch_files: schematic/kicad_playground.sch schematic/LEDs.sch
    board_name: kicad_playground

    output_dir: /cygdrive/c/Users/Stanley/Desktop/kdiff/kicad_playground/kidiff/schematic
       web_dir: /cygdrive/c/Users/Stanley/Desktop/kdiff/kicad_playground/kidiff/schematic/web/

git-kdiff-log master -- schematic/kicad_playground.kicad_pcb schematic/kicad_playground.sch schematic/LEDs.sch

Commits list:
cdb8005 | 2018-10-17 22:24:22 +0200 | Jean-Noël Avila | Add a subsheet
064acea | 2018-10-17 22:18:43 +0200 | Jean-Noël Avila | Fix Led direction
fc6818f | 2018-10-17 22:17:10 +0200 | Jean-Noël Avila | Small update to schematic
8b0696a | 2018-10-17 22:10:27 +0200 | Jean-Noël Avila | Draft of the design
c0a3b8e | 2018-10-17 21:57:38 +0200 | Jean-Noël Avila | First commit

Generating artifacts for 5 commits.
Totaling 20 possible comparisons.

Schematic Pages
     1  kicad_playground
     2  LEDs

Layers

   1. c0a3b8e - 8b0696a

kidiff schematic/kicad_playground.kicad_pcb --webserver-disable -a c0a3b8e -b 8b0696a &> /dev/null
Status: 127

plotgitsch -l schematic/kicad_playground-cache.lib -c FF0000:008FFF:FFFFFF:222222 c0a3b8e 8b0696a &> /dev/null
Status: 0

   2.         - fc6818f

kidiff schematic/kicad_playground.kicad_pcb --webserver-disable -a c0a3b8e -b fc6818f &> /dev/null
Status: 127

plotgitsch -l schematic/kicad_playground-cache.lib -c FF0000:008FFF:FFFFFF:222222 c0a3b8e fc6818f &> /dev/null
Status: 0

   3.         - 064acea

kidiff schematic/kicad_playground.kicad_pcb --webserver-disable -a c0a3b8e -b 064acea &> /dev/null
Status: 127

plotgitsch -l schematic/kicad_playground-cache.lib -c FF0000:008FFF:FFFFFF:222222 c0a3b8e 064acea &> /dev/null
Status: 0

   4.         - cdb8005

kidiff schematic/kicad_playground.kicad_pcb --webserver-disable -a c0a3b8e -b cdb8005 &> /dev/null
Status: 127

plotgitsch -l schematic/kicad_playground-cache.lib -c FF0000:008FFF:FFFFFF:222222 c0a3b8e cdb8005 &> /dev/null
Status: 0


Launching /cygdrive/c/Users/Stanley/Desktop/kdiff/kicad_playground/kidiff/schematic/web//main.html

```
This is the output of the terminal

**Comment by leoheck** [comment]
Hm, did you install python dependencies?
I have this PR to fix this. https://github.com/Gasman2014/KiCad-Diff/pull/47

Install these 2 python packages and try one more time.
- pygubu==0.10.5
- python_dateutil==2.8.1

**Comment by jorlee99** [comment]
> Hm, did you install python dependencies?
> I have this PR to fix this. [Gasman2014/KiCad-Diff#47](https://github.com/Gasman2014/KiCad-Diff/pull/47)
> 
> Install these 2 python packages and try one more time.
> 
> * pygubu==0.10.5
> * python_dateutil==2.8.1

So I just installed those packages and unfortunately no luck same issue.

**Comment by leoheck** [comment]
Run this and share the output please
```
kidiff schematic/kicad_playground.kicad_pcb --webserver-disable -a c0a3b8e -b cdb8005
```

**Comment by leoheck** [comment]
Your board is inside a folder called schematic, is this right?
`schematic/kicad_playground.kicad_pcb`

**Comment by jorlee99** [comment]
> Your board is inside a folder called schematic, is this right?
> `schematic/kicad_playground.kicad_pcb`
yup

> Run this and share the output please
> 
> ```
> kidiff schematic/kicad_playground.kicad_pcb --webserver-disable -a c0a3b8e -b cdb8005
> ```
I changed the directory to inside schematic and ran the following
`$ kdiff kicad_playground.kicad_pcb  --webserver-disable -a c0a3b8e -b cdb8005`
 and got the following output
`The board cdb8005 does not exit`

**Comment by leoheck** [comment]
Yes, you have to pass 2 valid git revisions where you know boards differs. You dont have changes in the layout. At leas try to change something, then commit and then run the kdiff again.

You can also explore kidiff with `kdiff kicad_playground.kicad_pcb` for instance.

**Comment by jorlee99** [comment]
I removed a component and added a new commit so totaling to 6 commits now:
```
$ kdiff schematic/kicad_playground.kicad_pcb

Generating artifacts for 6 commits.
Totaling 30 possible comparisons.

   1. c0a3b8e - 8b0696a
   2.         - fc6818f
   3.         - 064acea
   4.         - cdb8005
   5.         - 0ae4cf0

Launching /cygdrive/c/Users/Stanley/Desktop/kdiff/kicad_playground/kidiff/schematic/web//main.html


```
And when doing your command of comparing two commits 
`$ kdiff kicad_playground.kicad_pcb  --webserver-disable -a 8b0696a -b fc6818f`
I still get the following 
`The board fc6818f does not exit`

It seems like it's trying the use the commit as the file name currently? I'm not sure.

**Comment by leoheck** [comment]
Please, do this 
> You can also explore kidiff with kdiff kicad_playground.kicad_pcb for instance.

**Comment by jorlee99** [comment]
A quick update, I have played around with my linux system and an idea had was to send the files of the kdiff folder to my windows system. This produces the following:
![image](https://user-images.githubusercontent.com/43526839/118168540-cee19780-b3dc-11eb-9610-f75ebda25df5.png)

I think this means when running the kdiff command through bash on a windows system something messes up correct? However, looking at the PCB layout:
![image](https://user-images.githubusercontent.com/43526839/118168735-06504400-b3dd-11eb-8ea9-e8112ec26dab.png)

somehow the file goes missing (as the console can't find the svg) and it creates some place holder image.

**Comment by leoheck** [comment]
cool, let me see the main.html to check if the file references are correct.

**Comment by jorlee99** [comment]
> cool, let me see the main.html to check if the file references are correct.

My mistake it seems I was using a repo that didn't have PCB changes it's all working now:
![image](https://user-images.githubusercontent.com/43526839/118171279-056ce180-b3e0-11eb-92c4-1910db089c70.png)

So the file system of the output of kdiff on linux is fine. It just seems as though running the kdiff command through bash on windows it does not function?

**Comment by leoheck** [comment]
![image](https://user-images.githubusercontent.com/1277920/118172833-33175200-b403-11eb-9964-89f8d8154a38.png)

**Comment by jorlee99** [comment]
My apologies. Here are their respective kdiff files (one from windows the other from linux).
[kidifflinux.zip](https://github.com/leoheck/kdiff/files/6474534/kidifflinux.zip)
[kidiffwin.zip](https://github.com/leoheck/kdiff/files/6474555/kidiffwin.zip)
There are a couple things that might be notable when running on the windows machine? 
Firstly being when running `source env.sh` for kdiff
it outputs `-bash$'\r': command not found`
and when running `source env.sh` for the KiCad-Diff
it outputs `readlink: unknown option -- b`

Both seem ok? As I am running the kdiff command on the board I get the following as an example
![image](https://user-images.githubusercontent.com/43526839/118174335-88dc0200-b3e3-11eb-80c9-01297206ac42.png)
As if the components svgs in my kicad are creating an error.

Apologies again.

**Comment by leoheck** [comment]
Yeah these scripts have to be tweaked a little bit.

**Comment by jorlee99** [comment]
> Yeah these scripts have to be tweaked a little bit.

Looking a bit more into it, I have tried the KiCad-Diff kidiff command in bash on windows and as stated above it returns 
`/usr/bin/env: ‘python3’: No such file or directory`
Whereas on linux it runs just fine.
So do you think it may not be the script but rather KiCad-Diff itself because I noticed in the file structure comparing the windows linux, the index.html is missing. Any ideas?

**Comment by leoheck** [comment]
I am am running this just fine on Windows now.
Please, delete your repo Kicad-Diff and clone it again.\
Also, both env.sh do not show any error.
Maybe it will be necessary to delete and clone kdiff too

**Comment by leoheck** [comment]
Kicad-Diff (`kidiff`) is working for me now, the `env.sh` was ok

This is the command line
![image](https://user-images.githubusercontent.com/1277920/118201557-20664280-b42e-11eb-820a-4b26683abf73.png)

This is the output of the Kicad-Diff (`kidiff`)
![image](https://user-images.githubusercontent.com/1277920/118201504-fc0a6600-b42d-11eb-822d-3c23598bbb82.png)

I am updating the README with some instructions. Check there if you need to do something else.

**Comment by leoheck** [comment]
`plotgitch` is not working for me yet.
to solve this, I just made an empty script called plotgitch and added it to the PATH.

Then kdiff is generating layout diffs, but it cannot load the page at the end. I have to update this to be a server instead like Kicad-Diff is. 

To solve this, just load the Kicad-diff right after running kdiff. Like this

```
# Generate project diffs/images
project> kdiff board.kicad_pcb

# Worakaround to launch the server
project> kidiff board.kicad_pcb
```

Then a page will load, and then you have to replace the `index.html` to `main.html`
This is the URL that you have to use 
```
http://127.0.0.1:8888/web/main.html
```

![image](https://user-images.githubusercontent.com/1277920/118201904-fcefc780-b42e-11eb-846c-8df7084e90d0.png)

**Comment by leoheck** [comment]
Hey, @jorlee99 I pushed a change where it uses a webserver to launch the page at the end. 
Could you try if it works for you?

**Comment by jorlee99** [comment]
> Hey, @jorlee99 I pushed a change where it uses a webserver to launch the page at the end.
> Could you try if it works for you?
TLDR; Webpage launches but I am still encountering problems

Hi @leoheck, that's amazing that it's working on windows that means there is hope. Yea so your webserver seems to be working? For me it produces the following webpage.
![image](https://user-images.githubusercontent.com/43526839/118301004-6b687000-b497-11eb-93df-a276be185075.png)

So the webpage launches and that is great but i think the kdiff and kidiff commands are still broken for me somehow. Knowing that it works for you I must be missing something.

To start all over following your readme and unfortunately I will go painstakingly in depth so bear with me.
Running `Invoke-WebRequest -Uri https://aka.ms/wsl-ubuntu-2004 -OutFile ~/Downloads/ubuntu-2004.zip` this zip contains nothing in it 
[ubuntu-2004.zip](https://github.com/leoheck/kdiff/files/6480047/ubuntu-2004.zip)
So I downloaded the following [version of ubuntu](https://www.microsoft.com/en-us/p/ubuntu-1804/9n9tngvndl3q) please let me know if it is the correct one if not could you direct me to another one?

Next, using the installed ubuntu 18.04 LTS version, I ran all your dependency commands in your read me and all of those installed just fine. Now we start encountering the real problems. So opam did not let me create a new compiler and the list of compilers only goes up to 4.07.0 with trunk versions of the 4.08.0. I used the compiler 4.07.0 when trying all the upcoming commands so that might be the source of the issue. When running your script `install.sh` one thing to note was that I did need to `opam depext conf-m4.1` and `opam depext plotkicadsch` and `opam install digestif` or else it would throw errors. The Last error I got was this wall of text

![image](https://user-images.githubusercontent.com/43526839/118304041-ff880680-b49a-11eb-9c1a-8cea381fa260.png)

So I can't get plotkicadsch installed. Moving on, I ran `source env.sh` for both kdiff and KiCad-Diff and ran their respective commands kdiff and kidiff these are the outputs I get.

`kidiff` command
![image](https://user-images.githubusercontent.com/43526839/118304315-5ee61680-b49b-11eb-9130-dcc9c1966993.png)

`kdiff` command
![image](https://user-images.githubusercontent.com/43526839/118304380-758c6d80-b49b-11eb-9a52-44b4c307430e.png)

so obviously with kdiff because plotkicadsch didnt get install it doesn't have plotgitsch. 

Any ideas how to progress from here?

Thanks for your help so far though we seem close! @leoheck

**Comment by jorlee99** [comment]
> Hey, @jorlee99 I pushed a change where it uses a webserver to launch the page at the end.
> Could you try if it works for you?

However, on Linux and it works just fine 
![image](https://user-images.githubusercontent.com/43526839/118305376-b9cc3d80-b49c-11eb-876b-b7afd0ab2ef1.png)

**Comment by leoheck** [comment]
Yeap, this webserver still has an issue loading the page. I will check how to improve this.
What works for me is to kind `kdiff` then it will load the webserver which will be loading forever... then click in the URL field. And hit enter one more time. It may work.

**Comment by jorlee99** [comment]
> Yeap, this webserver still has an issue loading the page. I will check how to improve this.
> What works for me is to kind `kdiff` then it will load the webserver which will be loading forever... then click in the URL field. And hit enter one more time. It may work.

The webserver seems to running fine on linux so obviously it's a windows thing. But it seems to work with you by chance what ubuntu version are you running for your WSL.

**Comment by jorlee99** [comment]
I have an update. I have it fully working on a windows machine now. The key thing I was missing was that **you can NOT just do opam init on first installation**. After some digging it turns out that WSL does not support sandboxing otherwise known as the bwrap command as referenced [here](https://github.com/ocaml/opam/issues/3505) a quick workaround is doing the [following](https://discuss.ocaml.org/t/help-with-set-up-opam-sandboxing-fails-on-windows-wsl/3456/3) but do read the warnings concerning the matter as well. After doing that and following your other steps it works fine.  Thanks so much! @leoheck 



> Yeap, this webserver still has an issue loading the page. I will check how to improve this.
> What works for me is to kind `kdiff` then it will load the webserver which will be loading forever... then click in the URL field. And hit enter one more time. It may work.

as for the webpage issue you are correct sometimes it loads sometimes it doesn't but opening the file from the file explorer seems to work.

**Comment by leoheck** [comment]
Good, thanks for this finding. I am going to test to update the README.md. 
Thank you for your help @jorlee99.

**Comment by leoheck** [comment]
Hey @jorlee99 could you test the last version of the code to see if it improve page loading for you?
It should work normally.

**Comment by jorlee99** [comment]
Hi @leoheck. Running the kdiff command I get the following webpage 
![image](https://user-images.githubusercontent.com/43526839/118387236-31e84f80-b5d2-11eb-81c2-94bf022a852d.png)

I can access the directories and it leads me there eventually if I click the web folder.

**Comment by leoheck** [comment]
What is the command line you are using? 
From where you are running this command?

**Comment by leoheck** [comment]
I improve this code a little bit. Please, check if it is better now.
I still would like to know where you are running `kdiff` to check if this is related to your issue.

**Comment by jorlee99** [comment]
>What is the command line you are using?
I am currently using the ubuntu web store command line
> From where you are running this command?
I like to put all my files in a folder at the same level as the .git folder
Ex: if i were to check a schematic it would be within folder/board.kicad_pcb
therefore the command would be kdiff folder/board.kicad_pcb
> Are you inside the project folder? 
Yes

And yes it is working again i get the following after running the command
![image](https://user-images.githubusercontent.com/43526839/118420225-31ec5c00-b673-11eb-9db1-511691a3daf5.png)

**Comment by leoheck** [comment]
Cool, it looks like the nested project is not working anymore. I will try to address that today since I have repos with a nested projects

**Comment by jorlee99** [comment]
I can start creating projects within projects if you would like me to test nested projects as well.

**Comment by leoheck** [comment]
Cool, thanks.

I think plotgitsh is not prepared for this yet. However, I pushed a workaround here. Do you want to test?

You can test like this:

1. Create a repo, and put 2 kicad projects inside. Lets say
```
my-multiproject-repo/
   project1
      board.sch
   project2
      board.sch
```  
2. Make some changes in both `board.sch`
3. Launch kdiff and check if the Root (schematic folder) shows the right pages. Before it was mixing histories from these schematics with the same name.

**Comment by leoheck** [comment]
Hey, I am going to close this. 
Start a new issue if you see problems, cool?
